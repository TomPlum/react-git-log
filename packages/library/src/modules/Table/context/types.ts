export interface TableContextBag {
  /**
   * A timestamp format string passed to DayJS
   * to format the timestamps of the commits
   * in the log table.
   *
   * @default ISO-8601
   */
  timestampFormat?: string
}