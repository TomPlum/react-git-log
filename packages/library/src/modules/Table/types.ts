import { CSSProperties } from 'react'

/**
 * The number of pixels that the header
 * row needs to be so that the first
 * data row lines up with the first row
 * on the graph. Assuming 0 row spacing.
 */
export const HEADER_ROW_HEIGHT = 47

/**
 * The number of pixels of margin at the
 * top of the table element that offsets
 * it enough to line-up correctly with
 * the graph
 */
export const TABLE_MARGIN_TOP = 12

export interface TableProps {
  /**
   * A timestamp format string passed to DayJS
   * to format the timestamps of the commits
   * in the log table.
   *
   * @default ISO-8601
   */
  timestampFormat?: string

  /**
   * A class name passed to the tables
   * wrapping container element.
   */
  className?: string

  /**
   * A React CSS styling object passed to
   * the various elements of the table.
   */
  styles?: GitLogTableStylingProps
}

export interface GitLogTableStylingProps {
  /**
   * A React CSS styling object passed to
   * the container element of the table.
   */
  table?: CSSProperties

  /**
   * A React CSS styling object passed to
   * the wrapping element around the table
   * headers.
   */
  thead?: CSSProperties

  /**
   * A React CSS styling object passed to
   * each table row element.
   */
  tr?: CSSProperties

  /**
   * A React CSS styling object passed to
   * each table data element.
   */
  td?: CSSProperties
}