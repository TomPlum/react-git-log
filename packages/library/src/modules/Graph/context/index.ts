export * from './types'
export * from './GraphContext'
export * from './useGraphContext'