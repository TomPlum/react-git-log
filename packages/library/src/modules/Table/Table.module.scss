.table {
  display: grid;
  grid-template-columns: minmax(350px, 4fr) minmax(100px, 1fr) 195px;

  .row {
    display: contents;
  }

  .header:first-child {
    padding-left: 20px;
  }

  .header {
    font-weight: 600;
  }

  >.thead >.tr >.th {
    text-align: left;
    padding-bottom: 22px;
    box-sizing: border-box;
  }
}