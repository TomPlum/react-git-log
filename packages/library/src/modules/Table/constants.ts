/**
 * The number of pixels that the header
 * row needs to be so that the first
 * data row lines up with the first row
 * on the graph. Assuming 0 row spacing.
 */
export const HEADER_ROW_HEIGHT = 47

/**
 * The number of pixels of margin at the
 * top of the table element that offsets
 * it enough to line-up correctly with
 * the graph
 */
export const TABLE_MARGIN_TOP = 12