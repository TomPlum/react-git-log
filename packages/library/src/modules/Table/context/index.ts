export * from './types'
export * from './TableContext'
export * from './useTableContext'