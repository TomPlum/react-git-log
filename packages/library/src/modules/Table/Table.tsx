import styles from './Table.module.scss'
import dayjs from 'dayjs'
import { useMemo } from 'react'
import { Commit } from 'types/Commit'
import { useTheme } from 'hooks/useTheme'
import advancedFormat from 'dayjs/plugin/advancedFormat'
import relativeTime from 'dayjs/plugin/relativeTime'
import { useGitContext } from 'context/GitContext'
import { usePlaceholderData } from 'modules/Graph/strategies/Grid/hooks/usePlaceholderData'
import { TableRow } from 'modules/Table/components/TableRow'
import { TableContainer } from 'modules/Table/components/TableContainer'
import { TableProps } from './types'
import { TableContext, TableContextBag } from './context'

dayjs.extend(advancedFormat)
dayjs.extend(relativeTime)

export const Table = ({
  row,
  className,
  styles: styleOverrides,
  timestampFormat = 'YYYY-MM-DD HH:mm:ss'
}: TableProps) => {
  const { textColour, } = useTheme()
  const { placeholderData } = usePlaceholderData()
  const { showHeaders, graphData, paging, indexCommit, isIndexVisible } = useGitContext()

  const tableData = useMemo<Commit[]>(() => {
    const data = graphData.commits.slice(paging?.startIndex, paging?.endIndex)

    if (isIndexVisible && indexCommit) {
      data.unshift(indexCommit)
    }

    return data
  }, [graphData.commits, indexCommit, paging?.endIndex, isIndexVisible, paging?.startIndex])

  const tableContextValue = useMemo<TableContextBag>(() => ({
    timestampFormat
  }), [timestampFormat])

  return (
    <TableContext value={tableContextValue}>
      <TableContainer
        row={row}
        className={className}
        rowQuantity={tableData.length}
        styleOverrides={styleOverrides?.table}
      >
        {showHeaders && (
          <div
            className={styles.head}
            id='react-git-log-table-head'
            style={styleOverrides?.thead}
            data-testid='react-git-log-table-head'
          >
            <div
              className={styles.header}
              style={{ color: textColour }}
              id='react-git-log-table-header-commit-message'
              data-testid='react-git-log-table-header-commit-message'
            >
              Commit message
            </div>

            <div
              className={styles.header}
              style={{ color: textColour }}
              id='react-git-log-table-header-author'
              data-testid='react-git-log-table-header-author'
            >
              Author
            </div>

            <div
              className={styles.header}
              style={{ color: textColour }}
              id='react-git-log-table-header-timestamp'
              data-testid='react-git-log-table-header-timestamp'
            >
              Timestamp
            </div>
          </div>
        )}

        {tableData.length == 0 && placeholderData.map(({ commit }, i) => (
          <TableRow
            index={i}
            isPlaceholder
            commit={commit}
            rowStyleOverrides={styleOverrides?.tr}
            dataStyleOverrides={styleOverrides?.td}
            key={`git-log-empty-table-row-${commit.hash}`}
            data-testid={`react-git-log-empty-table-row-${i}`}
          />
        ))}

        {tableData.map((commit, i) => (
          <TableRow
            row={row}
            index={i}
            commit={commit}
            rowStyleOverrides={styleOverrides?.tr}
            dataStyleOverrides={styleOverrides?.td}
            key={`git-log-table-row-${commit.hash}`}
            data-testid={`react-git-log-table-row-${i}`}
          />
        ))}
      </TableContainer>
    </TableContext>
  )
}