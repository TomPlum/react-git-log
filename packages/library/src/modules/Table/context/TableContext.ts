import { createContext } from 'react'
import { TableContextBag } from 'modules/Table/context/types'

export const TableContext = createContext<TableContextBag>({
  timestampFormat: 'YYYY-MM-DD HH:mm:ss'
})