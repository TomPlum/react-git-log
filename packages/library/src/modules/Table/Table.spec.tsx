import { render, screen } from '@testing-library/react'
import { gitLog } from 'test/elements/GitLog'
import { table } from 'test/elements/Table'
import { Table } from 'modules/Table/Table'
import * as gitContext from 'context/GitContext'
import { commit, gitContextBag, graphData } from 'test/stubs'

describe('Table', () => {
  it('should pass the given table class to the git log table element', () => {
    render(
      <Table
        className='styles.customLogTableClass'
      />
    )

    const gitLogTable = gitLog.table()
    expect(gitLogTable).toBeInTheDocument()
    expect(gitLogTable.className).toContain('styles.customLogTableClass')
  })

  it('should pass the given table style objects to the git log table element', () => {
    vi.spyOn(gitContext, 'useGitContext').mockReturnValue(gitContextBag({
      showHeaders: true
    }))
    
    render(
      <Table
        styles={{
          table: {
            background: 'blueviolet'
          },
          thead: {
            background: 'darkolivegreen'
          },
          td: {
            background: 'mediumvioletred'
          },
          tr: {
            background: 'lightgoldenrodyellow'
          }
        }}
      />
    )

    const gitLogTable = gitLog.table()
    expect(gitLogTable).toBeInTheDocument()

    expect(table.container()?.style.background).toBe('blueviolet')
    expect(table.head()?.style.background).toBe('darkolivegreen')
    expect(table.emptyRow({ row: 0 })?.style.background).toBe('lightgoldenrodyellow')
    expect(table.timestampData({ row: 0 })?.style.background).toBe('mediumvioletred')
    expect(table.commitMessageData({ row: 0 })?.style.background).toBe('mediumvioletred')
  })

  it('should pass the given timestamp format string to the commit date formatter', () => {
    vi.useFakeTimers()
    vi.setSystemTime(new Date(2025, 2, 25))

    vi.spyOn(gitContext, 'useGitContext').mockReturnValue(gitContextBag({
      paging: {
        startIndex: 0,
        endIndex: 2,
      },
      isIndexVisible: false,
      graphData: graphData({
        commits: [
          commit({
            committerDate: '2025-03-05 17:03:58 +0000'
          })
        ]
      })
    }))

    render(
      <Table
        timestampFormat='YYYY MM'
      />
    )

    expect(table.timestampData({ row: 0 })).toHaveTextContent('2025 03')

    vi.useRealTimers()
  })

  it('should render a custom table row if one is passed', () => {
    const selectedCommit = commit({
      hash: '1'
    })

    const previewedCommit = commit({
      hash: '2',
      parents: ['1']
    })

    vi.spyOn(gitContext, 'useGitContext').mockReturnValue(gitContextBag({
      isIndexVisible: false,
      paging: {
        startIndex: 0,
        endIndex: 2,
      },
      selectedCommit,
      previewedCommit,
      graphData: graphData({
        positions: new Map([
          ['1', [0, 1]],
          ['2', [1, 1]],
        ]),
        commits: [
          selectedCommit,
          previewedCommit
        ]
      })
    }))

    const { asFragment } = render(
      <Table
        row={({ commit, selected, backgroundColour, previewed }) => {
          return (
            <div data-testid={`custom-table-row-${commit.hash}`}>
              <p data-testid={`custom-message-${commit.hash}`}>
                {commit.message}
              </p>

              <p data-testid={`custom-selected-${commit.hash}`}>
                {selected ? 'yes' : 'no'}
              </p>

              <p data-testid={`custom-previewed-${commit.hash}`}>
                {previewed ? 'yes' : 'no'}
              </p>

              <p data-testid={`custom-bg-colour-${commit.hash}`}>
                {backgroundColour}
              </p>
            </div>
          )
        }}
      />
    )

    // Hash 1 is the selected commit
    expect(screen.getByTestId('custom-table-row-1')).toBeInTheDocument()
    expect(screen.getByTestId('custom-message-1')).toHaveTextContent(selectedCommit.message)
    expect(screen.getByTestId('custom-selected-1')).toHaveTextContent('yes')
    expect(screen.getByTestId('custom-previewed-1')).toHaveTextContent('no')
    expect(screen.getByTestId('custom-bg-colour-1')).toHaveTextContent('rgba(41, 121, 255, 0.15)')

    // Hash 2 is the previewed commit
    expect(screen.getByTestId('custom-table-row-2')).toBeInTheDocument()
    expect(screen.getByTestId('custom-message-2')).toHaveTextContent(previewedCommit.message)
    expect(screen.getByTestId('custom-selected-2')).toHaveTextContent('no')
    expect(screen.getByTestId('custom-previewed-2')).toHaveTextContent('yes')
    expect(screen.getByTestId('custom-bg-colour-2')).toHaveTextContent('rgba(231, 231, 231, 0.5)')

    expect(asFragment()).toMatchSnapshot()
  })
})