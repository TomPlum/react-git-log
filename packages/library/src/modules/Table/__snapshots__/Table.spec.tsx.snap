// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > should render a custom table row if one is passed 1`] = `
<DocumentFragment>
  <div
    data-testid="react-git-log-table"
    id="react-git-log-table"
    style="margin-top: 12px;"
  >
    <div
      class="head"
      data-testid="react-git-log-table-head"
      id="react-git-log-table-head"
    >
      <div
        class="header"
        data-testid="react-git-log-table-header-commit-message"
        id="react-git-log-table-header-commit-message"
        style="color: rgb(0, 0, 0);"
      >
        Commit message
      </div>
      <div
        class="header"
        data-testid="react-git-log-table-header-author"
        id="react-git-log-table-header-author"
        style="color: rgb(0, 0, 0);"
      >
        Author
      </div>
      <div
        class="header"
        data-testid="react-git-log-table-header-timestamp"
        id="react-git-log-table-header-timestamp"
        style="color: rgb(0, 0, 0);"
      >
        Timestamp
      </div>
    </div>
    <div
      data-testid="custom-table-row-1"
      id="react-git-log-table-row-0"
    >
      <p
        data-testid="custom-message-1"
      >
        feat(graph): example commit message
      </p>
      <p
        data-testid="custom-selected-1"
      >
        yes
      </p>
      <p
        data-testid="custom-previewed-1"
      >
        no
      </p>
      <p
        data-testid="custom-bg-colour-1"
      >
        rgba(41, 121, 255, 0.15)
      </p>
    </div>
    <div
      data-testid="custom-table-row-2"
      id="react-git-log-table-row-1"
    >
      <p
        data-testid="custom-message-2"
      >
        feat(graph): example commit message
      </p>
      <p
        data-testid="custom-selected-2"
      >
        no
      </p>
      <p
        data-testid="custom-previewed-2"
      >
        yes
      </p>
      <p
        data-testid="custom-bg-colour-2"
      >
        rgba(231, 231, 231, 0.5)
      </p>
    </div>
  </div>
</DocumentFragment>
`;
